cmd/aws-vpc-cni/main.go

```diff
@@ -0,0 +1,365 @@
+	if val, ok := os.LookupEnv(env); ok {
+		return val
+	}
+	return def
+}
+
+// zero values of structs with omitempty is not supported, Golang issue for - https://github.com/golang/go/issues/11939
+// cni/types DNS struct is missing omitstring
+// NetConfList describes an ordered list of networks.
+type NetConfList struct {
+	CNIVersion string `json:"cniVersion,omitempty"`
+
+	Name         string     `json:"name,omitempty"`
+	DisableCheck bool       `json:"disableCheck,omitempty"`
+	Plugins      []*NetConf `json:"plugins,omitempty"`
+}
+
+// NetConf stores the common network config for the CNI plugin
+type NetConf struct {
+	CNIVersion string `json:"cniVersion,omitempty"`
+
+	Name         string          `json:"name,omitempty"`
+	Type         string          `json:"type,omitempty"`
+	Capabilities map[string]bool `json:"capabilities,omitempty"`
+	IPAM         *IPAMConfig     `json:"ipam,omitempty"`
+	DNS          *types.DNS      `json:"dns,omitempty"`
+
+	RawPrevResult map[string]interface{} `json:"prevResult,omitempty"`
+	PrevResult    types.Result           `json:"-"`
+
+	// Interface inside container to create
+	IfName string `json:"ifName,omitempty"`
+
+	Enabled string `json:"enabled,,omitempty"`
+
+	// IP to use as SNAT target
+	NodeIP net.IP `json:"nodeIP,omitempty"`
+
+	VethPrefix string `json:"vethPrefix,omitempty"`
+
+	// MTU for eth0
+	MTU string `json:"mtu,omitempty"`
+
+	PluginLogFile string `json:"pluginLogFile,omitempty"`
+
+	PluginLogLevel string `json:"pluginLogLevel,omitempty"`
+}
+
+//Ref : https://github.com/containernetworking/plugins/plugins/ipam/host-local/backend/allocator/config.go
+type IPAMConfig struct {
+	*Range
+	Name       string         `json:"name,omitempty"`
+	Type       string         `json:"type,omitempty"`
+	Routes     []*types.Route `json:"routes,omitempty"`
+	DataDir    string         `json:"dataDir,omitempty"`
+	ResolvConf string         `json:"resolvConf,omitempty"`
+	Ranges     []RangeSet     `json:"ranges"`
+	IPArgs     []net.IP       `json:"-"` // Requested IPs from CNI_ARGS and args
+}
+
+type RangeSet []Range
+
+type Range struct {
+	RangeStart net.IP      `json:"rangeStart,omitempty"` // The first ip, inclusive
+	RangeEnd   net.IP      `json:"rangeEnd,omitempty"`   // The last ip, inclusive
+	Subnet     types.IPNet `json:"subnet"`
+	Gateway    net.IP      `json:"gateway,omitempty"`
+}
+
+func waitForIPAM() bool {
+	for {
+		cmd := exec.Command("./grpc-health-probe", "-addr", "127.0.0.1:50051", ">", "/dev/null", "2>&1")
+		var outb bytes.Buffer
+		cmd.Stdout = &outb
+		cmd.Run()
+		if outb.String() == "" {
+			return true
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func getNodePrimaryV4Address() (string, error) {
+	var hostIP string
+	var err error
+	for {
+		hostIP, err = imds.GetMetaData("local-ipv4")
+		if err != nil {
+			log.WithError(err).Fatalf("aws-vpc-cni init failed\n")
+			return "", err
+		}
+		if hostIP != "" {
+			return hostIP, nil
+		}
+
+		time.Sleep(1 * time.Second)
```

Comment:
nit: seems like `GetMetaData` already has a timeout built in, so do we need this sleep?

Reply:
So there are 10 retries so after that I added this sleep for the next retry. But let me check if we can remove it, we can save a second.